### DISCUSSION TITLE
关于部分书不能刷的解决。

### DISCUSSION THREAD

**Discussion by C260210** [Mufanc/iSmartAuto#45] [Ideas]
在实际遇到书籍的时候，遇到了bookType为1的书籍，但修改bookType后发现仍然是没有运行效果
出现的问题是如下的：
2022-05-30 02:05:21.997 | SUCCESS | automaton.spider.spider:book_info:214 - [获取书籍信息] | 获取书籍信息完成
2022-05-30 02:05:21.998 | INFO | automaton.spider.spider:get_tasks:118 - [构建任务树] | 正在构建任务树...
2022-05-30 02:05:22.097 | SUCCESS | automaton.spider.spider:get_tasks:140 - [构建任务树] | 构建任务树完成
2022-05-30 02:05:22.097 | SUCCESS | automaton.spider.spider:get_tasks:141 - [获取任务列表] | 获取任务列表完成
2022-05-30 02:05:22.098 | INFO | automaton.utils:dfs:78 - [刷任务点] | 大学体验英语视听说2
2022-05-30 02:05:22.099 | INFO | automaton.utils:dfs:78 - [刷任务点] | . Unit 1 Knowledge is Great
2022-05-30 02:05:22.100 | INFO | automaton.utils:dfs:78 - [刷任务点] | . Lead-in
2022-05-30 02:05:22.101 | INFO | automaton.utils:dfs:78 - [刷任务点] | . Lead-in
......中间省略一大段相同的......
2022-05-30 02:05:22.165 | INFO | automaton.utils:dfs:78 - [刷任务点] | . Video Studio
2022-05-30 02:05:22.165 | INFO | automaton.utils:dfs:78 - [刷任务点] | . Unit 6 All on the Mind
都是直接跳过，通过分析，

```

> 
> > async def _flash(course_id, book_id, spider):
> > try:
> > async def dfs(node, depth=0):
> > task = node.task
> > logger.info(f"[刷任务点] | {PLACEHOLDER * depth} {task['name']}")
> > if _paper_config['skip-finished'] and task['unitStudyPercent'] == 100:
> > logger.info(f"[刷任务点] | {PLACEHOLDER * depth} 跳过")
> > return
> > logger.success("success2")
> > if 'paperId' in task:  # 如果有任务则提交
> > chapter_id = task['chapterId']
> > task_id = task['id']
> > score, time = await _random(spider, task['paperId'])
> > result = await page.submit(book_id, chapter_id, task_id, score, time, 100, user_id)
> > if result['wasThrown'] or not result['result']['value']:
> > logger.warning(f'[刷任务点] | 任务[{task["name"]}]可能提交失败，请留意最终结果')
> > for ch in node.child:
> > await dfs(ch, depth + 1)
> > 
> > if 'paperId' in task:  问题是出现在这一段上，这类书籍的task中并没有paperId的，然后一步步分析可以知道paperId的作用是在get_paper函数中获取答案的，而获取完答案后再返回到 _random(）函数。
> > 其实这一步可以直接跳过，观察可知get_paper()函数最后返回到的是 _random(）函数，而 _random()函数的作用就是随机生成分数和时间，那为什么不直接生成分数和时间呢？
> > 
> > 由此我们可以直接将代码修改为 如下：
> > 

```

async def _flash(course_id, book_id, spider):
    try:
        async def dfs(node, depth=0):
            task = node.task
            logger.info(f"[刷任务点] | {PLACEHOLDER * depth} {task['name']}")
            if _paper_config['skip-finished'] and task['unitStudyPercent'] == 100:
                logger.info(f"[刷任务点] | {PLACEHOLDER * depth} 跳过")
                return
            logger.success("success2")
            if 'id' in task and task["id"] != '':   # 如果有任务则提交
                chapter_id = task['chapterId']
                task_id = task['id']
                score, time = 100,50 #直接指定分数
                result = await page.submit(book_id, chapter_id, task_id, score, time, 100, user_id)
                if result['wasThrown'] or not result['result']['value']:
                    logger.warning(f'[刷任务点] | 任务[{task["name"]}]可能提交失败，请留意最终结果')
            for ch in node.child:
                await dfs(ch, depth + 1)

这样就可以直接上传数据，实际也确实上传了数据。但事实确实进度和分数还是没有变化
那是不是说不同类型的书籍的上传的信息也是有所不同的呢？
接下来我们进行抓包，通过抓包我们可以看到果然上传的信息中是没有result这个参数的，而是一个叫qstJson的东西，那接下来
我们只需要将result改为qstJson即可
[{
	"uid": "1996928",
	"taskNo": "1",
	"taskId": "1468101c1da99f2aeab65422a1390c5f",
	"studyDate": "1672122091688",
	"deviceType": 5,
	"percent": "100",
	"batchTag": "85DF3DE47B65400DB65DBF11592BC788",
	"seconds": "312",
	"chapterId": "22564bdc66d94648affeb89e78f1ebef",
	"id": "3726",
	"qstJson": [{
		"elementId": "bdc8622db95bd690439c9ffd66820033",
		"order": "11",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":11,\"question_type\":\"8\",\"user_score\":\"0\",\"result\":2,\"user_answer\":\"assertive\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "98aa655515525ff242e990cdf5d49700",
		"order": "12",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":12,\"question_type\":\"8\",\"user_score\":\"0\",\"result\":2,\"user_answer\":\"cheerful\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "76f17e70cb3d42dc5cda6e2eb38ca9fd",
		"order": "5",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":5,\"question_type\":\"8\",\"user_score\":\"0\",\"result\":2,\"user_answer\":\"outgoing\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "b0042a60c4d42098daa665c3b1b3fc8c",
		"order": "7",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":7,\"question_type\":\"8\",\"user_score\":\"0\",\"result\":2,\"user_answer\":\"kind-hearted\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "9f8722e586acd69867988958845b89aa",
		"order": "3",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":3,\"question_type\":\"8\",\"user_score\":\"1\",\"result\":1,\"user_answer\":\"dependable\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "7fb4c134ba10e1b4458f2f4b5b92aaa1",
		"order": "0",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091688,
		"elementType": 1,
		"elementAttr": "{\"order\":0,\"question_type\":\"8\",\"user_score\":\"3\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "3ec378f0bb0f9aa066556ca612cecc3a",
		"order": "1",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":1,\"question_type\":\"8\",\"user_score\":\"1\",\"result\":1,\"user_answer\":\"honest\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "0c4e7d078a602d846bf16b9935ef9ea9",
		"order": "6",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":6,\"question_type\":\"8\",\"user_score\":\"0\",\"result\":2,\"user_answer\":\"friendly\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "2f492f61bfa7c65a42c471d8730113e9",
		"order": "10",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":10,\"question_type\":\"8\",\"user_score\":\"0\",\"result\":2,\"user_answer\":\"mature\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "d1ace061999df7f1bae1270a15f46ff9",
		"order": "8",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":8,\"question_type\":\"8\",\"user_score\":\"0\",\"result\":2,\"user_answer\":\"easygoing\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "0f324388be87a1ed6c0b9bad0a008f0c",
		"order": "4",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":4,\"question_type\":\"8\",\"user_score\":\"0\",\"result\":2,\"user_answer\":\"shy\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "2c1b44b7b652a5aac4853d3073ca02c8",
		"order": "2",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":2,\"question_type\":\"8\",\"user_score\":\"1\",\"result\":1,\"user_answer\":\"sincere\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}, {
		"elementId": "ccc9758b3de975ae03864468251c5908",
		"order": "9",
		"uid": "1996928",
		"taskId": "1468101c1da99f2aeab65422a1390c5f",
		"deviceType": 5,
		"batchTag": "7AAFDE838C50457F8E2EB30D5C0F3C14",
		"pageId": "b1e00218c6db4a899c214ee3dbe467ea",
		"chapterId": "22564bdc66d94648affeb89e78f1ebef",
		"addTime": 1672122091672,
		"elementType": 2,
		"elementAttr": "{\"order\":9,\"question_type\":\"8\",\"user_score\":\"0\",\"result\":2,\"user_answer\":\"easygoing\",\"question_id\":\"a43b457b65714abb8b672c0e5a432237\"}",
		"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
	}],
	"score": "25",
	"bookId": "d0f8f7a090b74d7ca97fdf22066ad622"
}]

**Comment by aried202** [comment]
具体修改哪一段啊

**Comment by TWLW9784** [comment]
最后一步在哪里更改啊？

**Comment by cwwk** [comment]
result数据在哪里修改a，在utils里直接替换还是没有上传数据

**Reply by puxuhui1123** [reply]
请问现在解决了没有
